### PR TITLE
FW: condition builtin testlist on CPU arch

### PR DESCRIPTION
**Pull request by mcisowsk**

### CODE REVIEW

File: framework/test_selectors/SelectorFactory.cpp
Comment (mcisowsk) [inline framework/test_selectors/SelectorFactory.cpp:193]: from CodingGuidelines:
Constants and defines should be ALL_CAPS.

File: framework/scripts/generate_test_list.py
Comment (jposwiata) [inline framework/scripts/generate_test_list.py:91]: if no default list exists, this will cause SEG_FAULTS. Add appropriate if before compares.

File: framework/test_selectors/SelectorFactory.cpp
Comment (jposwiata) [inline framework/test_selectors/SelectorFactory.cpp:229]: seems that "sometimes" nullptr might be passed in the list. Add appropriate check to avoid SEGFAULTS.

**Review by jposwiata — CHANGES_REQUESTED** [review]
Seems that the way of "detection" doesn't suit our needs. Would cpu features be better?

File: framework/test_selectors/SelectorFactory.cpp
Comment (thiagomacieira) [reply framework/test_selectors/SelectorFactory.cpp:193]: I'd prefer we use `ALL_CAPS` only for macros. This is not one.

File: framework/test_selectors/SelectorFactory.cpp
Comment (thiagomacieira) [inline framework/test_selectors/SelectorFactory.cpp:214]: Please use CPU ISA features to detect the CPU, not the model-stepping.

There's code in `dump_cpu_info()`  in sandstone.cpp to try and detect this, but it's currently broken and only detecting Core2.

**Review by thiagomacieira — CHANGES_REQUESTED** [review]

File: framework/test_selectors/SelectorFactory.cpp
Comment (jposwiata) [inline framework/test_selectors/SelectorFactory.cpp:200]: tolower()? (in simd.conf "short names" are uppercase, I think dictionaries are lowercase..)
Could you attach snippet of generated cpu_features.h?

**Review by jposwiata — CHANGES_REQUESTED** [review]

File: framework/test_selectors/SelectorFactory.cpp
Comment (mcisowsk) [reply framework/test_selectors/SelectorFactory.cpp:200]: `static const struct X86Architecture x86_architectures[] = {
    { cpu_core2, "Core2", "x86_64" },
    { cpu_westmere, "Westmere", "wsm" },
    { cpu_sandybridge, "Sandy Bridge", "snb" },
    { cpu_silvermont, "Silvermont", "slm" },
    { cpu_ivybridge, "Ivy Bridge", "ivb" },
    { cpu_goldmont, "Goldmont", "glm" },
    { cpu_haswell, "Haswell", "hsw" },
    { cpu_broadwell, "Broadwell", "bdw" },
    { cpu_tremont, "Tremont", "tnt" },
    { cpu_skylake, "Skylake", "skl" },
    { cpu_skylake_avx512, "Skylake (Avx512)", "skx" },
    { cpu_cascadelake, "Cascade Lake", "clx" },
    { cpu_cooperlake, "Cooper Lake", "cpx" },`

Comment (mcisowsk) [reply framework/test_selectors/SelectorFactory.cpp:200]: I used lc() in perl script

Comment (jposwiata) [reply framework/test_selectors/SelectorFactory.cpp:200]: I've missed it :D

File: framework/cpuid_internal.h
Comment (thiagomacieira) [reply framework/cpuid_internal.h:222]: That should be the last one that matches. That's why they are sorted. You just need to iterate backwards, actually.

Comment (thiagomacieira) [reply framework/cpuid_internal.h:222]: #457

Comment (mcisowsk) [reply framework/cpuid_internal.h:222]: Ok I will rebase on it once it is merged

Comment (jposwiata) [reply framework/cpuid_internal.h:222]: If the list is in "reverse" order the comment has no sense.

File: framework/scripts/generate_test_list.py
Comment (mcisowsk) [inline framework/scripts/generate_test_list.py:100]: to check

Comment (mcisowsk) [reply framework/scripts/generate_test_list.py:100]: to do it like in detect_cpu_arch(), that is with masking `(cpu_features & arch) == arch`, it would require from user to define the lists in `-Dbuiltin_test_list=default:tests/data/test_list_default.csv,icx:tests/data/test_list_icx.csv...`  in a correct order. This is rather insider knowledge, do we want to require user to know that?

Comment (mcisowsk) [reply framework/scripts/generate_test_list.py:100]: for example, core2 testlist should go as the last one

File: framework/scripts/generate_test_list.py
Comment (jposwiata) [inline framework/scripts/generate_test_list.py:110]: use lambda here.

Comment (mcisowsk) [reply framework/scripts/generate_test_list.py:110]: oh, right. Thanks

**Review by jposwiata — DISMISSED** [review]
LGTM

File: framework/scripts/generate_test_list.py
Comment (thiagomacieira) [inline framework/scripts/generate_test_list.py:101]: This sorting may produce false positives because the CPU features aren't added always at the end in the CPUID and our bits are therefore also all out of order.  Take for example `cpu_feature_avxvnni`, which is higher than `cpu_feature_avx512f`. That means a list for Sierra Forest would have this bit set and its numeric value would be higher than a list for Ice Lake. Then, when running on a CPU that has both features (e.g., Sapphire Rapids), we may select the SRF one instead of the ICX one.

One way around this is to sort by bit population count instead of numeric value (use [`std::popcount`](https://en.cppreference.com/w/cpp/numeric/popcount)). It may still fail, but is less likely.

Another is to not sort at all (except maybe the "default" fallback last) but instead require the maintainers to pass the lists in the proper search order.

File: framework/cpuid_internal.h
Comment (thiagomacieira) [inline framework/cpuid_internal.h:260]: You didn't use any of these modifications or the ones in main.cpp

File: framework/sandstone.cpp
Comment (thiagomacieira) [inline framework/sandstone.cpp:1292]: You've accidentally removed this in rebasing (from #457)

File: framework/test_selectors/SelectorFactory.cpp
Comment (thiagomacieira) [inline framework/test_selectors/SelectorFactory.cpp:210]: I don't think this is the right place to put this. I would prefer for it to be in sandstone.cpp's `main()`, where we're already calling `get_test_lsit()` (actually, there may be no need for modifications, at all, if you make the function handle a `nullptr` name be the same as "auto").

**Review by thiagomacieira — CHANGES_REQUESTED** [review]

File: framework/scripts/generate_test_list.py
Comment (mcisowsk) [reply framework/scripts/generate_test_list.py:101]: no sorting then

File: framework/test_selectors/SelectorFactory.cpp
Comment (mcisowsk) [inline framework/test_selectors/SelectorFactory.cpp:213]: If I do not use this flag correctly, then it will be better to create separate file for this specific function auto_detect_test_list(), so that it won't be included in the unittests source files

File: framework/scripts/generate_test_list.py
Comment (thiagomacieira) [inline framework/scripts/generate_test_list.py:92]: Same problem as before: this is a static non-POD, so it runs too early. How about a plain C array?

This is unfortunately not a theoretical or academic exercise. Looking at the generated code, I see:
```
Disassembly of section .text.startup:

0000000000000000 <_GLOBAL__sub_I_sandstone_test_lists.cpp>:
   0:   55                      push   %rbp
   1:   48 8d 05 00 00 00 00    lea    0x0(%rip),%rax        # 8 <_GLOBAL__sub_I_sandstone_test_lists.cpp+0x8>
                        4: R_X86_64_PC32        .data.rel.ro.local-0x4
   8:   48 8d 0d 00 00 00 00    lea    0x0(%rip),%rcx        # f <_GLOBAL__sub_I_sandstone_test_lists.cpp+0xf>
                        b: R_X86_64_PC32        .data.rel.ro.local+0x5c
   f:   c4 e1 f9 6e c1          vmovq  %rcx,%xmm0
```
There's an AVX instruction there.

File: framework/test_selectors/SelectorFactory.cpp
Comment (thiagomacieira) [inline framework/test_selectors/SelectorFactory.cpp:210]: Pass a `std::span` here instead of `std::vector`

File: framework/test_selectors/SelectorFactory.cpp
Comment (thiagomacieira) [reply framework/test_selectors/SelectorFactory.cpp:213]: That's ok, the same happens above (line 83)

I actually prefer to have it in this file, instead of generated,  because we do need to apply the sorting. The current generated code for the internal product has:
```c++
    &icx_test_list,
    &cpx_test_list,
    &skx_test_list,
    &spr_test_list,
```

This is not in any special sorting order (ICX later than CPX and SKX, but earlier than SPR). So we must either sort in this function (at a minimum to `assert` that the supplied order is correct).

**Review by thiagomacieira — CHANGES_REQUESTED** [review]
Nearly there.

File: framework/test_selectors/SelectorFactory.cpp
Comment (mcisowsk) [reply framework/test_selectors/SelectorFactory.cpp:213]: But I thought you agreed that there can be no sorting at all, as no method ensures 100% correct sorting. User should define the architecture lists in the correct order.

File: framework/scripts/generate_test_list.py
Comment (thiagomacieira) [inline framework/scripts/generate_test_list.py:101]: This is creating an array of pointers:
```c++
static const struct TestList* test_lists[] = {
    &icx_test_list,
    &cpx_test_list,
    &skx_test_list,
    &spr_test_list,
    &default_test_list
};
```

So there's one indirection too many: we could have a simple array of `TestList` items.

File: framework/test_selectors/SelectorFactory.cpp
Comment (thiagomacieira) [reply framework/test_selectors/SelectorFactory.cpp:213]: We can go with that. If all lists are supersets of previous ones, we could search for the one that matches most bits... but the ISA extensions weren't a monotonic sequence. So I agree on enforcing the order on whoever supplies the lists.

**Review by thiagomacieira — APPROVED** [review]

File: framework/scripts/generate_test_list.py
Comment (jposwiata) [reply framework/scripts/generate_test_list.py:101]: This list is in incorrect order (the order from the option had been agreed, selector factory will use default list in most cases).

**Review by jposwiata — CHANGES_REQUESTED** [review]
The per-gen list is in incorrect order (sorted with increased number of features).

**Comment by thiagomacieira** [comment]
Superseded by #486 .

**Pull request closed without merging**
